src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs

```diff
@@ -216,10 +216,11 @@ public void SaveTo(IHostEnvironment env, Stream outputStream)
 
         IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
 
-        public IRowToRowMapper GetRowToRowMapper(DataViewSchema inputSchema)
+        IRowToRowMapper ITransformer.GetRowToRowMapper(DataViewSchema inputSchema)
         {
             Contracts.CheckValue(inputSchema, nameof(inputSchema));
-            Contracts.Check(IsRowToRowMapper, nameof(GetRowToRowMapper) + " method called despite " + nameof(IsRowToRowMapper) + " being false.");
+            Contracts.Check(((ITransformer)this).IsRowToRowMapper, nameof(ITransformer.GetRowToRowMapper) + " method called despite " +
```

Comment:
>IsRowToRowMapper [](start = 49, length = 16)

I feel we should hide `IsRowToRowMapper` as well?

Reply:
I think you are asking whether we can make `IsRowToRowMapper` internal. 

I discussed this with Tom at length and the answer is that we should not make it internal. The reason is that when a `ITransformer` is a `IRowToRowMapper`, the performance of the prediction engine can be significantly increased compared to when it is not. Tom talked of a difference of orders of magnitude. 

Since it is such an important characteristics of the `ITransformer` we think it is better to keep it as part of the public API. This is especially true when we think about possibly allowing expansion of the library with custom components in the future.

Reply:
If its for potential use, I feel we should hide for this version because I don't think users will be able to control the performance of ML.NET in the near future.

---
In reply to: [259442579](https://github.com/dotnet/machinelearning/pull/2673#discussion_r259442579) [](ancestors = 259442579)

Reply:
I will let @TomFinley explain his reasons.

Reply:
I think in general the policy is a good one, that we should try to limit the scope of the public API. However, I feel very strongly that something like `IRowToRowMapper` has to exist.

The reason why I think this is the case is that I think back to #973. Literally everything *can* be captured as mapping from `IDataView` to `IDataView`, as we have seen. But we see that if we have a more "specific" interface for this row-to-row mapping, we can gain some benefit -- in this case, orders of magnitude speedup and a far more clear interface from using this "bypass" mechanism, in the case where we *can* use it.

From this perspective, we might say, "very well, then `IRowToRowMapper` should just be internal and prediction engine will still be free to use it." Which is of course possible. But then I think to myself, if `PredictionEngine` is useful from a reflection-driven perspective (as it quite clearly is), what is the dynamic API's answer to this scenario? There is nothing except `IRowToRowMapper`.

So, I think the case for something like `IRowToRowMapper` is clear, so I think we should ship it. So does it have to be that specifically? I don't see any better proposal than that interface. (Though perhaps it could be refined, e.g., with @sfilipi making it operating over schema columns specifically rather than `int => bool` delegates, etc.) But I think the principle has proven itself useful and should be made public, just like `IDataView` has been around, proven itself useful, and should be made public.

Reply:
(I know this already closed, but though I'd answer the direct question. 😄 )